pkg/proctree/proctree_feed.go

```diff
@@ -0,0 +1,246 @@
+	thread.GetInfo().SetFeedAt(
+		TaskInfoFeed{
+			Name:        leader.GetInfo().GetName(),
+			Tid:         int(feed.ChildTid),
+			Pid:         int(feed.ChildPid),
+			NsTid:       int(feed.ChildNsTid),
+			NsPid:       int(feed.ChildNsPid),
+			StartTimeNS: feed.ChildStartTime,
+			PPid:        int(feed.ParentPid),
+			NsPPid:      int(feed.ParentNsPid),
+			Uid:         0, // TODO: implement
+			Gid:         0, // TODO: implement
+		},
+		utils.NsSinceBootTimeToTime(feed.TimeStamp),
+	)
+
+	thread.SetParentHash(feed.ParentHash) // all threads have the same parent as the thread group leader
+	thread.SetLeaderHash(feed.LeaderHash) // thread group leader is a "process" (not a thread)
+	leader.AddThread(feed.ChildHash)      // add the thread to the thread group leader
+
+	return nil
+}
+
+type ExecFeed struct {
+	TimeStamp         uint64
+	TaskHash          uint32
+	ParentHash        uint32
+	LeaderHash        uint32
+	CmdPath           string
+	PathName          string
+	Dev               uint32
+	Inode             uint64
+	Ctime             uint64
+	InodeMode         uint16
+	InterPathName     string
+	InterDev          uint32
+	InterInode        uint64
+	InterCtime        uint64
+	Interpreter       string
+	StdinType         uint16
+	StdinPath         string
+	InvokedFromKernel int32
+}
+
+// FeedFromExec feeds the process tree with an exec event.
+func (pt *ProcessTree) FeedFromExec(feed ExecFeed) error {
+	if feed.TaskHash != feed.LeaderHash {
+		// Running execve() from a thread is discouraged and behavior can be unexpected:
+		//
+		// 1. All threads are terminated.
+		// 2. PID remains the same.
+		// 3. New process is single threaded.
+		// 4. Inherited Attributed (open fds, proc group, ID, UID, GID, ... are retained)
+		// 5. Still, it isn't forbidden and should be handled.
+
+		// TODO: handle execve() from a thread
+		logger.Debugw("exec event received for a thread", "taskHash", feed.TaskHash)
+		return nil
+	}
+
+	// Update the process: it is very likely that the process already exists because of the fork
+	// event. There are small chances that it might not exist yet due to signal event ordering. If
+	// that is the case, the process will be created and only the information about the
+	// executable/interpreter will be updated. This way, when the fork event is received, the
+	// process will be updated with the correct information.
+
+	process := pt.GetOrCreateProcessByHash(feed.TaskHash) // create if not exists
+
+	process.SetParentHash(feed.ParentHash) // faster than checking if already set
+
+	process.GetInfo().SetNameAt(
+		feed.CmdPath,
+		utils.NsSinceBootTimeToTime(feed.TimeStamp),
+	)
+
+	process.GetExecutable().SetFeedAt(
+		FileInfoFeed{
+			Name:      feed.CmdPath,
+			Path:      feed.PathName,
+			Dev:       int(feed.Dev),
+			Ctime:     int(feed.Ctime),
+			Inode:     int(feed.Inode),
+			InodeMode: int(feed.InodeMode),
+		},
+		utils.NsSinceBootTimeToTime(feed.TimeStamp),
+	)
+
+	process.GetInterpreter().SetFeedAt(
```

Comment:
You have an error here.
There are 2 types of interpreters, and you mish-mashed both of them.
One is the kernel "interpreter", which is the loader of the binary. Most of the times its `ld.so`. This is a binary which is loaded to the memory and its entry-point is called before the executed binary entry-point is called.
The other type of interpreter is what we call "interp". This is the binary which really runs, when a non-elf file is run. For example, when running shell script which starts with '#!', the binary executed is the binary. However, the real binary run is the interpreter program, for example bash. So the "interp" is the shell interpreter in this case.
As you can see, we can have 3 different files when running a program. If we execute "run.sh", then the binary is "run.sh", the interp is "/bin/bash" and the interpreter is "/lib/ld.so".

Now about the mistake - you took for the interpreter file the information of both "interp" and "interpreter".
The names you gave them are also misleading - what is "inter" and what is "interpreter"? Our current names are already much too confusing.

You should probably have all 3 binaries saved for a task. I think you can extract the "interp" file information from the `bprm->executed` field, but it depends on the kernel version and doing so is out of scope for this PR.
Other option is to have the "interp" as some field of the executed file data.

Reply:
Yep, I always saw the "interp" as being different from other values of interpreter, but always left it "to check later". Let me try to tackle this, lets see...

Reply:
BTW I agree Interpreter is a terrible name for the "bin format loader". It should be called "BinFormatLoader" or "BinaryInterpreter", or something else.